src/buildlog/mod.rs

```diff
@@ -28,10 +30,68 @@ impl BuildLog {
 		BuildLog { entries: BTreeMap::new() }
 	}
 
+	pub fn estimated_total_task_time(&self, output: RawString, _command: RawString) -> Option<Duration> {
+		match self.entries.get(&output) {
+			Some(entry) => Some(Duration::from_millis((entry.end_time_ms - entry.start_time_ms).into())),
+			None => {
+				// TODO: Search entries for command_hash equal to murmur_hash_64a(_command)
+				None
+			},
+		}
+	}
+
+	pub fn average_historic_task_time(&self) -> Option<Duration> {
+		if self.entries.is_empty() {
+			None
+		} else {
+			let mut sum_ms = 0 as u64;
+			for (_output, entry) in self.entries.iter() {
+				sum_ms += (entry.end_time_ms - entry.start_time_ms) as u64;
+			}
+			Some(Duration::from_millis(sum_ms / self.entries.len() as u64))
+		}
+	}
+
+	pub fn add_entry(&mut self, output: RawString, build_starttime: Instant, starttime: Instant, endtime: Instant, command: RawString) {
+		self.entries.insert(output, Entry {
+			start_time_ms: (starttime - build_starttime).as_millis() as u32,
+			end_time_ms:   (endtime - build_starttime).as_millis() as u32,
+			restat_mtime:  0 /* TODO */,
+			command_hash:  murmur_hash_64a(command.as_bytes()),
+		});
+	}
+
 	pub fn read(file: impl AsRef<Path>) -> Result<BuildLog, Error> {
 		BuildLog::read_from(File::open(file)?)
 	}
 
+	pub fn write(&self, file: impl AsRef<Path>) -> Result<(), Error> {
+		// TODO: should we append to this file instead of truncating it?
+		self.write_to(File::create(file)?)
+	}
+
+	pub fn write_to(&self, file: File) -> Result<(), Error> {
+		let mut file = BufWriter::new(file);
+
+		file.write(b"# ninja log v5\n")?;
+
+		// Write entries in order of finishing time. Note that this removes all
+		// 'dead' entries immediately, while ninja would only do that later.
+
+		let mut entries = Vec::from_iter(&self.entries);
```

Comment:
Instead of copying and sorting here, it might make sense to use a different data structure which preserves the order of added entries. [`index`](https://docs.rs/indexmap/1.0.2/indexmap/) might be useful.

Or just append the new entries directly to the file, instead of re-writing the file at the end.